modules/core/src/main/java/org/apache/ignite/spi/deployment/local/LocalDeploymentSpi.java

```diff
@@ -171,38 +237,33 @@ private String getResourceName(String clsName, Map<String, String> rsrcs) {
     @Override public boolean register(ClassLoader ldr, Class<?> rsrc) throws IgniteSpiException {
         assert ldr != null;
         assert rsrc != null;
+        boolean move = false;
 
         if (log.isDebugEnabled())
             log.debug("Registering [ldrRsrcs=" + ldrRsrcs + ", ldr=" + ldr + ", rsrc=" + rsrc + ']');
 
-        ConcurrentMap<String, String> clsLdrRsrcs = ldrRsrcs.getSafe(ldr);
+        Map<String, String> newRsrcs;
 
-        if (clsLdrRsrcs == null) {
-            ConcurrentMap<String, String> old = ldrRsrcs.putIfAbsent(ldr,
-                clsLdrRsrcs = new ConcurrentHashMap<>());
-
-            if (old != null)
-                clsLdrRsrcs = old;
-        }
+        synchronized (this) {
+            ConcurrentMap<String, String> clsLdrRsrcs = ldrRsrcs.getSafe(ldr);
 
-        Map<String, String> newRsrcs = addResource(ldr, clsLdrRsrcs, rsrc);
+            // move forward, localDeployTask compatibility issue.
+            if (clsLdrRsrcs != null && ldrRsrcs.size() > 1) {
+                ldrRsrcs.remove(ldr);
 
-        if (!getBoolean(IGNITE_DEPLOYMENT_PRESERVE_LOCAL)) {
-            Collection<ClassLoader> rmvClsLdrs = null;
+                move = true;
+            }
 
-            if (!F.isEmpty(newRsrcs)) {
-                rmvClsLdrs = new LinkedList<>();
+            ConcurrentMap<String, String> old = ldrRsrcs.putIfAbsent(ldr,
+                clsLdrRsrcs == null ? clsLdrRsrcs = new ConcurrentLinkedHashMap<>() : clsLdrRsrcs);
 
-                removeResources(ldr, newRsrcs, rmvClsLdrs);
-            }
+            if (old != null)
+                clsLdrRsrcs = old;
 
-            if (rmvClsLdrs != null) {
-                for (ClassLoader cldLdr : rmvClsLdrs)
-                    onClassLoaderReleased(cldLdr);
-            }
+            newRsrcs = addResource(ldr, clsLdrRsrcs, rsrc);
         }
 
-        return !F.isEmpty(newRsrcs);
+        return !F.isEmpty(newRsrcs) && !move;
```

Comment:
This is a suspecting expression.
I think you should stay it as before (!F.isEmpty(newRsrcs)).

Reply:
done